File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/claims/MyClaimsListViewModel.kt
@file:Suppress("PackageNaming", "MagicNumber")

package app.pantopus.android.ui.screens.homes.claims

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.homes.OwnershipClaimDto
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.homes.HomesRepository
import app.pantopus.android.ui.components.StatusChipVariant
import app.pantopus.android.ui.screens.shared.list_of_rows.ListOfRowsUiState
import app.pantopus.android.ui.screens.shared.list_of_rows.RowLeading
import app.pantopus.android.ui.screens.shared.list_of_rows.RowModel
import app.pantopus.android.ui.screens.shared.list_of_rows.RowSection
import app.pantopus.android.ui.screens.shared.list_of_rows.RowTemplate
import app.pantopus.android.ui.screens.shared.list_of_rows.RowTrailing
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import timber.log.Timber
import java.time.Instant
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import javax.inject.Inject

/**
 * ViewModel for "My claims" list — wraps
 * `GET /api/homes/my-ownership-claims`. Backend masks the internal
 * state to a generic `status` string; we map the three known values to
 * chip variants below.
 *
 * TODO(design): claim-status detail screen not yet designed — taps
 * currently log only.
 */
@HiltViewModel
class MyClaimsListViewModel
    @Inject
    constructor(
        private val repo: HomesRepository,
    ) : ViewModel() {
        private val _state = MutableStateFlow<ListOfRowsUiState>(ListOfRowsUiState.Loading)
        val state: StateFlow<ListOfRowsUiState> = _state.asStateFlow()

        private var onStartNewClaim: () -> Unit = {}

        fun configureNavigation(onStartNewClaim: () -> Unit) {
            this.onStartNewClaim = onStartNewClaim
        }

        fun load() {
            if (_state.value is ListOfRowsUiState.Loaded) return
            refresh()
        }

        fun refresh() {
            _state.value = ListOfRowsUiState.Loading
            viewModelScope.launch {
                when (val result = repo.myOwnershipClaims()) {
                    is NetworkResult.Success -> applySuccess(result.data.claims)
                    is NetworkResult.Failure ->
                        _state.value = ListOfRowsUiState.Error(result.error.message)
                }
            }
        }

        private fun applySuccess(claims: List<OwnershipClaimDto>) {
            if (claims.isEmpty()) {
                _state.value =
                    ListOfRowsUiState.Empty(
                        icon = PantopusIcon.ShieldCheck,
                        headline = "No claims yet",
                        subcopy = "Once you submit a claim, you'll see its status here.",
                        ctaTitle = "Claim a home",
                        onCta = onStartNewClaim,
                    )
                return
            }
            val rows = claims.map(::rowFor)
            _state.value =
                ListOfRowsUiState.Loaded(
                    sections = listOf(RowSection(id = "my-claims", rows = rows)),
                    hasMore = false,
                )
        }

        private fun rowFor(claim: OwnershipClaimDto): RowModel =
            RowModel(
                id = claim.id,
                title = "Claim ${claim.id.take(8)}",
                subtitle = subtitleFor(claim),
                template = RowTemplate.StatusChip,
                leading = RowLeading.Icon(PantopusIcon.ShieldCheck, PantopusColors.primary600),
                trailing =
                    RowTrailing.Status(
                        text = statusText(claim.status),
                        variant = statusVariant(claim.status),
                    ),
                onTap = { Timber.tag("MyClaimsList").i("Claim row tapped: %s", claim.id) },
            )

        private fun subtitleFor(claim: OwnershipClaimDto): String? {
            val parts =
                listOfNotNull(
                    friendlyMethod(claim.method),
                    relativeSubmitted(claim.createdAt),
                )
            return parts.takeIf { it.isNotEmpty() }?.joinToString(" · ")
        }

        private fun friendlyMethod(method: String): String? =
            when (method) {
                "doc_upload" -> "Document upload"
                "fast_track" -> "Fast-track invite"
                "id_verification" -> "ID verification"
                else -> method.takeIf { it.isNotEmpty() }
            }

        private fun relativeSubmitted(iso: String): String? =
            runCatching {
                val instant = DateTimeFormatter.ISO_OFFSET_DATE_TIME.parse(iso, Instant::from)
                val days = ChronoUnit.DAYS.between(instant, Instant.now())
                val hours = ChronoUnit.HOURS.between(instant, Instant.now())
                val minutes = ChronoUnit.MINUTES.between(instant, Instant.now())
                "Submitted " +
                    when {
                        days >= 7L -> "${days / 7L}w ago"
                        days >= 1L -> "${days}d ago"
                        hours >= 1L -> "${hours}h ago"
                        minutes >= 1L -> "${minutes}m ago"
                        else -> "just now"
                    }
            }.getOrNull()

        private fun statusText(status: String): String =
            when (status) {
                "verified", "approved", "complete" -> "Verified"
                "rejected", "denied" -> "Not approved"
                "under_review", "pending", "submitted" -> "Under review"
                else -> status.replace('_', ' ').replaceFirstChar(Char::uppercase)
            }

        private fun statusVariant(status: String): StatusChipVariant =
            when (status) {
                "verified", "approved", "complete" -> StatusChipVariant.Success
                "rejected", "denied" -> StatusChipVariant.ErrorVariant
                "under_review", "pending", "submitted" -> StatusChipVariant.Info
                else -> StatusChipVariant.Neutral
            }
    }

File: frontend/apps/ios/Pantopus/Features/Homes/Claims/MyClaimsListViewModel.swift
//
//  MyClaimsListViewModel.swift
//  Pantopus
//
//  Backs `MyClaimsListView`. Fetches `GET /api/homes/my-ownership-claims`
//  (route `backend/routes/homeOwnership.js:217`) and maps each claim to
//  a `status_chip` row. Backend always returns the opaque
//  `"under_review"` status while a claim is open; once approved /
//  rejected the same field flips. We map those three values to chip
//  variants below.
//
//  TODO(design): a richer claim-status detail view isn't drawn yet —
//  tapping a row currently no-ops with a console log.
//

import Foundation
import Logging
import Observation
import SwiftUI

/// ViewModel for the "My claims" list.
@Observable
@MainActor
final class MyClaimsListViewModel: ListOfRowsDataSource {
    let title = "My claims"
    var topBarAction: TopBarAction? { nil }
    let tabs: [ListOfRowsTab] = []
    var selectedTab: String = ""
    var fab: FABAction? { nil }

    private(set) var state: ListOfRowsState = .loading

    private let api: APIClient
    private let onStartNewClaim: @Sendable () -> Void
    private let logger = Logger(label: "app.pantopus.ios.MyClaimsList")

    init(
        api: APIClient = .shared,
        onStartNewClaim: @escaping @Sendable () -> Void = {}
    ) {
        self.api = api
        self.onStartNewClaim = onStartNewClaim
    }

    func load() async {
        // Unlike MyHomes, we always refetch — claim status changes
        // server-side within hours (per the success-step copy) and the
        // user expects to see status flips on return-nav, not only via
        // pull-to-refresh.
        if case .loading = state {} else {
            state = .loading
        }
        await fetch()
    }

    func refresh() async {
        await fetch()
    }

    /// Endpoint isn't paginated server-side.
    func loadMoreIfNeeded() async {}

    private func fetch() async {
        do {
            let response: MyOwnershipClaimsResponse =
                try await api.request(HomesEndpoints.myOwnershipClaims())
            if response.claims.isEmpty {
                state = .empty(
                    ListOfRowsState.EmptyContent(
                        icon: .shieldCheck,
                        headline: "No claims yet",
                        // Empty-state CTA opens the AddHome wizard
                        // (which kicks off verification when the user
                        // selects "Owner" on the role step). The "Add a
                        // home" copy matches the wizard the CTA actually
                        // routes to, so the user doesn't expect a
                        // claim-existing-home picker that doesn't exist.
                        subcopy: "Submit a claim from a home dashboard. New here? Add a home and pick the Owner role to start.",
                        ctaTitle: "Add a home",
                        onCTA: onStartNewClaim
                    )
                )
            } else {
                let rows = response.claims.map { row(for: $0) }
                state = .loaded(sections: [RowSection(rows: rows)], hasMore: false)
            }
        } catch {
            state = .error(
                message: (error as? APIError)?.errorDescription
                    ?? "Couldn't load your claims."
            )
        }
    }

    private func row(for claim: OwnershipClaimDTO) -> RowModel {
        RowModel(
            id: claim.id,
            title: "Claim \(claim.id.prefix(8))",
            subtitle: subtitle(for: claim),
            template: .statusChip,
            leading: .icon(.shieldCheck, tint: Theme.Color.primary600),
            trailing: .statusChip(
                text: statusText(for: claim.status),
                variant: statusVariant(for: claim.status)
            ),
            onTap: { @Sendable [logger] in
                Task { @MainActor in
                    logger.info("Claim row tapped: \(claim.id) — detail view not yet designed.")
                }
            }
        )
    }

    private func subtitle(for claim: OwnershipClaimDTO) -> String? {
        // Show the friendlier method + the relative submitted date.
        let method = friendlyMethod(claim.method)
        let relative = relativeDate(claim.createdAt)
        return [method, relative].compactMap { $0 }.joined(separator: " · ")
    }

    private func friendlyMethod(_ method: String) -> String? {
        switch method {
        case "doc_upload": return "Document upload"
        case "fast_track": return "Fast-track invite"
        case "id_verification": return "ID verification"
        default: return method.isEmpty ? nil : method
        }
    }

    private func relativeDate(_ iso: String) -> String? {
        let parser = ISO8601DateFormatter()
        parser.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        guard let date = parser.date(from: iso) ?? ISO8601DateFormatter().date(from: iso) else {
            return nil
        }
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .short
        return "Submitted \(formatter.localizedString(for: date, relativeTo: Date()))"
    }

    private func statusText(for status: String) -> String {
        switch status {
        case "verified", "approved", "complete": return "Verified"
        case "rejected", "denied": return "Not approved"
        case "under_review", "pending", "submitted": return "Under review"
        default: return status.replacingOccurrences(of: "_", with: " ").capitalized
        }
    }

    private func statusVariant(for status: String) -> StatusChipVariant {
        switch status {
        case "verified", "approved", "complete": return .success
        case "rejected", "denied": return .error
        case "under_review", "pending", "submitted": return .info
        default: return .neutral
        }
    }
}
